File: src/main/java/aoc2019/day17/Day17.java
package aoc2019.day17;

import aoc2019.day09.Output;

import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static aoc2019.day09.Day9Kt.processParameterMode;

public class Day17 {

    static Map<Point, String> processInput(ArrayList<Long> numbers) {
        int i = 0;
        int relativeBase = 0;
        Map<Point, String> map = new HashMap<>();
        int x = 0, y = 0;

        while (i < numbers.size()) {
            int opCode = Math.toIntExact(numbers.get(i));
            if (opCode == 99) {
                break;
            }
            Output output = processParameterMode(numbers, i, opCode, 0, relativeBase);

            if (output.getCode().length() > 0) {
                switch (output.getCode()) {
                    case "35":
                        map.put(new Point(x++, y), "#");
                        break;
                    case "46":
                        map.put(new Point(x++, y), ".");
                        break;
                    case "10":
                        x = 0;
                        y++;
                }
            }

            if (output.getRelativeBase() != 0) {
                relativeBase = output.getRelativeBase();
            }

            i += output.getIndex();
        }
        return map;
    }

    static void printMap(Map<Point, String> map) {
        int maxX = map.keySet().stream().mapToInt(p -> p.x).max().orElse(-1);
        int maxY = map.keySet().stream().mapToInt(p -> p.y).max().orElse(-1);

        for (int y = 0; y < maxY; y++) {
            for (int x = 0; x < maxX; x++) {
                String p = map.get(new Point(x, y));
                System.out.print(p != null ? p : " ");
            }
            System.out.println();
        }
    }

    static java.util.List<Point> getIntersections(Map<Point, String> map) {
        return map.entrySet().stream().filter(entry -> entry.getValue().equals("#"))
                .map(Map.Entry::getKey)
                .filter(current -> "#".equals(map.get(new Point(current.x, current.y - 1))) &&
                        "#".equals(map.get(new Point(current.x, current.y + 1))) &&
                        "#".equals(map.get(new Point(current.x - 1, current.y))) &&
                        "#".equals(map.get(new Point(current.x + 1, current.y)))).collect(Collectors.toList());
    }

    static int multiplyCoordinates(List<Point> list) {
        return list.stream().mapToInt(p -> p.x * p.y).sum();
    }

}

File: src/main/kotlin/aoc2019/day17/Day17.kt
package aoc2019.day17

import aoc2019.commons.IntCodeProgram
import aoc2019.day09.readInput
import aoc2020.day20.Coord
import util.readInputLineByLine

fun readInput(path: String): List<Long> {
    return readInputLineByLine(path).map { it.toLong() }
}

fun computeIntersections(input: List<Long>): Int {
    val output = IntCodeProgram(input)
    output.execute()
    val outputMap = asciiListToMap(output)
    val intersections = getIntersections(outputMap)
    return multiplyCoordinatesAndSum(intersections)
}

private fun asciiListToMap(output: IntCodeProgram): MutableMap<Coord, Char> {
    val outputMap = mutableMapOf<Coord, Char>()
    var x = 0
    var y = 0
    for (char in output.outputToAscii())
        if (char == '\n') {
            y++
            x = 0
        } else {
            outputMap[Coord(x, y)] = char
            x++
        }
    return outputMap
}


fun processInputPart2(input: List<Long>, putValAtIndex: Pair<Long, Int>): Map<Coord, Char> {
    val output = IntCodeProgram(input)
    output.execute()
    val outputMap = asciiListToMap(output)
    return outputMap
}

fun printMap(map: Map<Coord, Char>) {
    val maxX = map.keys.stream().mapToInt { p: Coord -> p.x }.max().orElse(-1)
    val maxY = map.keys.stream().mapToInt { p: Coord -> p.y }.max().orElse(-1)
    for (y in 0..maxY)
        for (x in 0..maxX) {
            val p = map[Coord(x, y)]
            print(p ?: " ")
        }
}

fun getIntersections(map: Map<Coord, Char>): List<Coord> {
    return map.entries
        .filter { it.value == '#' }
        .map { it.key }
        .filter { current: Coord ->
            '#' == map[Coord(current.x, current.y - 1)]
                    && '#' == map[Coord(current.x, current.y + 1)]
                    && '#' == map[Coord(current.x - 1, current.y)]
                    && '#' == map[Coord(current.x + 1, current.y)]
        }
}

fun multiplyCoordinatesAndSum(list: List<Coord>): Int {
    return list.map { p: Coord -> p.x * p.y }.sum()
}

fun main() {
    val input = readInput("src/main/kotlin/aoc2019/day17/input1")
    input[0] = 2
    val output = IntCodeProgram(input)
    output.execute()
    printMap(asciiListToMap(output))
}